### PR TITLE
ExoPlayer 2.17.1 update and MediaSource management rework

### PR DESCRIPTION
**Pull request by karyogamy**
#### What is it?
- [ ] Bugfix (user facing)
- [x] Feature (user facing)
- [ ] Codebase improvement (dev facing)
- [x] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
Hi team,
I took a shot at upgrading ExoPlayer to 2.17.0 and realized that quite a few API has been made obsolete. For example, `getTag` has been dropped and will need even more hacks to get it to work. Since a rework is needed anyways, I took the liberty to toss out the old custom MediaSource management in favour of a more vanilla ExoPlayer model, following  [this comment](https://github.com/google/ExoPlayer/issues/7087#issuecomment-599581609). This is an attempt to help with refactoring the player so we can shift to expose ExoPlayer interface for player management instead of managing our own `PlayQueue` for playlist modifications.

This means ExoPlayer will now be responsible for managing the internal playlist, and `PlayQueue` will no longer play a central role to playback management. It will become a player proxy for user issued commands, and acts as a view for the player playlist, since ExoPlayer doesn't provide an API for the entire playlist. As an added bonus, shuffle now works properly. You may also notice buffer will start immediately when seeking to an unprepared stream, as opposed to the old model, where playback resumes until `StreamInfo` is fetched before buffering starts. This is because `StreamInfo` are now loaded on-demand inside the `MediaSource` in a custom `CompositeMediaSource` that dispatches a call to resolve/build the actual `MediaSource` using the `PlaybackResolvers`. 

Also, each `MediaSource` is now associated with a `MediaItem` that holds metadata. `Tag` objects are now contained inside this `MediaItem` instead. As a result, the old `MediaSourceTag` are now revamped into a generic interface that bridges between `PlayQueueItem` and `StreamInfo` since they both hold similar metadata. This can be expanded later so this interface object is exposed to external adapters and properly display if a stream is properly loaded or has errors (e.g. content error).

Going back to `MediaSource` management, a challenge arising from the new model is its lifecycle, where `MediaSource` in ExoPlayer aren't really allowed to self update. This means when a stream expires, we cannot update the `Playlist MediaSource` with a new one, nor should we since ExoPlayer should take care of it instead. This brings in `DataSpec` resolution, which is an interface ExoPlayer exposes that lets us examine an URI and replace it with another. So the solution here is to update our `PlaybackResolver` to encapsulate the Extractor Stream metadata (e.g. `AudioStream`, `VideoStream`, `SubtitleStream`, etc.) in an URI, and deliver that to the new `DataSpecResolver` to convert our custom URI to a real stream URI. This means we can rely on our own Extractor cache to handle expiration and resolve stream metadata into the real URL and ExoPlayer will handle the rest.

Concerns:
1. Because ExoPlayer now takes care of managing its own playlist, the `MediaSourceManager` (now `PlayQueueReactor`) no longer correct desynchronizations between the `PlayQueue` and ExoPlayer's own playlist. Since `PlayQueue` and `PlayQueueReactor` communicate with a message bus, non-atomic calls to `PlayQueue` can cause desyncs. For example, `Enqueue next` was broken because it works by appending to the queue and moving the last item to the current playing. When the append event message gets to the reactor, the item has already moved on the play queue. This is temporarily fixed by introducing a new event `Insert` to avoid mixing logic with auto-queuing on `Append` events. But a proper solution would be to refactor `PlayQueue` into an interface and create a new wrapper `ActivePlayQueue` on the base `PlayQueue` to handle all the player controlled play queue logic. The ultimate end-state for `PlayQueue` in this new model is to mainly act as a quick-access view for the adapters, and all actions to the queue are exposed via ExoPlayer interface instead.

2. Since ExoPlayer controls the shuffling now, we also lose some control (for the time being) in shuffled mode. For example, currently `PlayQueue` shuffle reorders the queue and is presented to the user (I actually like this better) instead of ExoPlayer's model of jumping to a random index. But this now also breaks the `Enqueue next` feature when used in shuffle mode. We can potentially create our own shuffle order as an individual component, though it means more code and kind of defeats the purpose of lifting the feature to ExoPlayer.

3. As mentioned before, `StreamInfo` preparation will now also be part of the buffering. This means, when selecting a stream that is not yet loadedm at all, buffering will feel a bit slower than before. Whereas before, the `StreamInfo` would load in the background while the stream continues to play before the buffering starts. I tried to mitigate this by pre-emptively loading the `StreamInfo` of the next `MediaItem`, but this issue will never fully go away since we can't load every item on the queue.

As you can see, there are quite a few more things that can be worked on to get this to a perfect state. But since the PR size is getting larger, I wanted to get this PR out now to get some opinion on if we want to make this shift in player modelling as part of the greater refactor work (or if this path isn't really worth proceeding). And if we want to make the shift, have the changes tested by more people and the code reviewed since it might break existing features.

#### APK testing 

The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR.

#### Fixes the following issue(s)
- relate to https://github.com/TeamNewPipe/NewPipe/issues/7939

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

### CODE REVIEW

File: app/src/main/java/org/schabi/newpipe/player/Player.java [hunk 1]
```diff
@@ -270,25 +271,26 @@
...
     @Nullable private PlayQueueItem currentItem;
-    @Nullable private MediaSourceTag currentMetadata;
+    @Nullable private MediaItemTag currentMetadata;
     @Nullable private Bitmap currentThumbnail;
 
     /*//////////////////////////////////////////////////////////////////////////
     // Player
     //////////////////////////////////////////////////////////////////////////*/
 
-    private SimpleExoPlayer simpleExoPlayer;
```
Comment (karyogamy) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:283]: `SimpleExoPlayer` is being deprecated.

File: app/src/main/java/org/schabi/newpipe/player/Player.java [hunk 2]
```diff
@@ -270,25 +271,26 @@
...
+    private ExoPlayer simpleExoPlayer;
     private AudioReactor audioReactor;
     private MediaSessionManager mediaSessionManager;
     @Nullable private SurfaceHolderCallback surfaceHolderCallback;
 
-    @NonNull private final CustomTrackSelector trackSelector;
```
Comment (karyogamy) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:288]: `CustomTrackSelector` interface has conflicts now, `DefaultTrackSelector` already fixed the issue with language selection, so making the switch here.

File: app/src/main/java/org/schabi/newpipe/player/Player.java [hunk 3]
```diff
@@ -1980,17 +1983,39 @@ private void hideSystemUIIfNeeded() {
     //////////////////////////////////////////////////////////////////////////*/
     //region Playback states
 
-    @Override // exoplayer listener
-    public void onPlayerStateChanged(final boolean playWhenReady, final int playbackState) {
```
Comment (karyogamy) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:1984]: `onPlayerStateChanged` interface is now broken into separate callbacks below.

File: app/src/main/java/org/schabi/newpipe/player/Player.java [hunk 4]
```diff
@@ -4498,9 +4469,13 @@ private void setupVideoSurface() {
             surfaceHolderCallback = new SurfaceHolderCallback(context, simpleExoPlayer);
             binding.surfaceView.getHolder().addCallback(surfaceHolderCallback);
             final Surface surface = binding.surfaceView.getHolder().getSurface();
-            // initially set the surface manually otherwise
-            // onRenderedFirstFrame() will not be called
-            simpleExoPlayer.setVideoSurface(surface);
+            // ensure player is using an unreleased surface, which the surfaceView might not be
+            // when starting playback on background or during player switching
+            if (surface.isValid()) {
```
Comment (karyogamy) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:4474]: @TiA4f8R This should fix the issue with starting livestream on background. Let me know if this works and if you want this in another PR instead.

File: app/src/main/java/org/schabi/newpipe/player/helper/CacheFactory.java [hunk 5]
```diff
@@ -49,7 +49,9 @@ private CacheFactory(@NonNull final Context context,
                          final long maxFileSize) {
         this.maxFileSize = maxFileSize;
 
-        dataSourceFactory = new DefaultDataSourceFactory(context, userAgent, transferListener);
```
Comment (karyogamy) [inline app/src/main/java/org/schabi/newpipe/player/helper/CacheFactory.java:52]: More ExoPlayer deprecation.

File: app/src/main/java/org/schabi/newpipe/player/helper/MediaSessionManager.java [hunk 6]
```diff
@@ -19,7 +19,6 @@
 import org.schabi.newpipe.MainActivity;
 import org.schabi.newpipe.player.mediasession.MediaSessionCallback;
 import org.schabi.newpipe.player.mediasession.PlayQueueNavigator;
-import org.schabi.newpipe.player.mediasession.PlayQueuePlaybackController;
```
Comment (karyogamy) [inline app/src/main/java/org/schabi/newpipe/player/helper/MediaSessionManager.java:22]: Deprecated and removed from Exo.

File: app/src/main/java/org/schabi/newpipe/player/helper/PlayerDataSource.java [hunk 7]
```diff
@@ -80,17 +83,6 @@ public PlayerDataSource(@NonNull final Context context, @NonNull final String us
         return new DefaultDashChunkSource.Factory(dataSourceFactory);
     }
 
-    public HlsMediaSource.Factory getHlsMediaSourceFactory() {
```
Comment (karyogamy) [inline app/src/main/java/org/schabi/newpipe/player/helper/PlayerDataSource.java:83]: Removed because I don't think there is caching for livestreams. Let me know if otherwise to revert.

File: app/src/main/java/org/schabi/newpipe/player/mediasession/PlayQueueNavigator.java [hunk 8]
```diff
@@ -44,17 +43,17 @@ public long getSupportedQueueNavigatorActions(@Nullable final Player player) {
     }
 
     @Override
-    public void onTimelineChanged(final Player player) {
+    public void onTimelineChanged(@NonNull final Player player) {
```
Comment (karyogamy) [inline app/src/main/java/org/schabi/newpipe/player/mediasession/PlayQueueNavigator.java:46]: More deprecations here.

File: app/src/main/java/org/schabi/newpipe/player/resolver/ExtractorDataSourceResolver.java [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+public final class ExtractorDataSourceResolver implements ResolvingDataSource.Resolver {
```
Comment (litetex) [inline app/src/main/java/org/schabi/newpipe/player/resolver/ExtractorDataSourceResolver.java:32]: I'm not sure if 3 interfaces and 4 classes inside that class are a good idea. Maybe create separate classes?

Comment (litetex) [inline app/src/main/java/org/schabi/newpipe/player/resolver/ExtractorDataSourceResolver.java:32]: Why is this method final?

**Review by litetex — CHANGES_REQUESTED** [review]
Thank you for the PR!

A ExoPlayer update was long overdue in my opinion, good that you picked it up.

Very good PR in my opinion, I like the coding style and the usage of new Java 8 features such as the Streams API 👍 

Didn't test it yet but I found some things in the review that should be processed:

Update: Also please have a look at the Sonarcloud warnings

File: app/src/main/java/org/schabi/newpipe/player/resolver/ExtractorDataSourceResolver.java [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+        public Uri resourceUriFrom(@NonNull final StreamInfo info, @NonNull final String protocol) {
+            final Uri manifestUri;
+            if (DASH_PROTOCOL.equals(protocol) && info.getDashMpdUrl() != null) {
+                manifestUri = Uri.parse(info.getDashMpdUrl());
+            } else if (HLS_PROTOCOL.equals(protocol) && info.getHlsUrl() != null) {
+                manifestUri = Uri.parse(info.getHlsUrl());
+            } else {
+                throw new NoSuchElementException("Protocol [" + protocol + "] not found.");
+            }
+            return manifestUri;
```
[the remark this answers is not in the document]
Comment (Redirion) [reply app/src/main/java/org/schabi/newpipe/player/resolver/ExtractorDataSourceResolver.java:299]: subjective. I actually prefer this coding style with preferably one return statement or an Exception (which is done here) or at least as few as possible return statements if the code would otherwise get ugly.

[the remark this answers is not in the document]
Comment (litetex) [reply app/src/main/java/org/schabi/newpipe/player/resolver/ExtractorDataSourceResolver.java:299]: Sorry but I think that 
```java
            if (DASH_PROTOCOL.equals(protocol) && info.getDashMpdUrl() != null) {
                return Uri.parse(info.getDashMpdUrl());
            } else if (HLS_PROTOCOL.equals(protocol) && info.getHlsUrl() != null) {
                return Uri.parse(info.getHlsUrl());
            }
            throw new NoSuchElementException("Protocol [" + protocol + "] not found.");
```
is a lot better readable then:
```java
            final Uri manifestUri;
            if (DASH_PROTOCOL.equals(protocol) && info.getDashMpdUrl() != null) {
                manifestUri = Uri.parse(info.getDashMpdUrl());
            } else if (HLS_PROTOCOL.equals(protocol) && info.getHlsUrl() != null) {
                manifestUri = Uri.parse(info.getHlsUrl());
            } else {
                throw new NoSuchElementException("Protocol [" + protocol + "] not found.");
            }
            return manifestUri;
```

Shorter, less expressions, less ``{ }``, same functionality.

**Comment by karyogamy** [comment]
Many thanks for the thorough review @litetex. 

Given the serialization usage coincides with the `DataSpec` issue that might conflict with @TiA4f8R's `delivery-methods-v2`, I'd recommend not going forward with this PR as-is, and instead continue with the existing model where we have control `MediaSource`. The player changes though will largely remain the same, so the review is still effective. I've acked most of the change requests, and will make the change and add docs when time permits. 

Eventually when we decide to move to the ExoPlayer model, we will have the same issue using `DataSpec`s and need to choose to go with serialization or cache.

File: app/src/main/java/org/schabi/newpipe/player/Player.java [hunk 11]
```diff
@@ -1950,10 +1962,12 @@ private void showOrHideButtons() {
         final boolean showNext = playQueue.getIndex() + 1 != playQueue.getStreams().size();
         final boolean showQueue = playQueue.getStreams().size() > 1 && !popupPlayerSelected();
         boolean showSegment = false;
-        if (currentMetadata != null) {
-            showSegment = !currentMetadata.getMetadata().getStreamSegments().isEmpty()
-                    && !popupPlayerSelected();
-        }
+        showSegment = /*only when stream has segment and playing in fullscreen player*/
```
Comment (litetex) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:1965]: Maybe also do that in the line above?

Comment (litetex) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:1965]:
```suggestion
        showSegment = /*only when stream has segments and is not playing in popup player*/
```

File: app/src/main/java/org/schabi/newpipe/player/Player.java [hunk 12]
```diff
@@ -1993,9 +2007,30 @@ private void hideSystemUIIfNeeded() {
...
     //////////////////////////////////////////////////////////////////////////*/
     //region Playback states
+    @Override
+    public void onPlayWhenReadyChanged(final boolean playWhenReady, final int reason) {
+        if (DEBUG) {
+            Log.d(TAG, "ExoPlayer - onPlayWhenReadyChanged() called with: "
+                    + "playWhenReady = [" + playWhenReady + "], "
+                    + "reason = [" + reason + "]");
+        }
+        final int playbackState = simpleExoPlayer == null
```
Comment (litetex) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:2017]: Use `exoPlayerIsNull()`

File: app/src/main/java/org/schabi/newpipe/player/Player.java [hunk 13]
```diff
@@ -1993,9 +2007,30 @@ private void hideSystemUIIfNeeded() {
...
 
-    @Override // exoplayer listener
-    public void onPlayerStateChanged(final boolean playWhenReady, final int playbackState) {
+    @Override
+    public void onPlaybackStateChanged(final int playbackState) {
+        if (DEBUG) {
+            Log.d(TAG, "ExoPlayer - onPlaybackStateChanged() called with: "
+                    + "playbackState = [" + playbackState + "]");
+        }
+        final boolean playWhenReady = simpleExoPlayer != null && simpleExoPlayer.getPlayWhenReady();
```
Comment (litetex) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:2029]: Use `getPlayWhenReady()` here (no variable required then)

File: app/src/main/java/org/schabi/newpipe/player/Player.java [hunk 14]
```diff
@@ -2461,26 +2487,37 @@ private void setMuteButton(@NonNull final ImageButton button, final boolean isMu
...
-        if (DEBUG) {
-            Log.d(TAG, "ExoPlayer - onTimelineChanged() called with "
-                    + "timeline size = [" + timeline.getWindowCount() + "], "
-                    + "reason = [" + reason + "]");
-        }
-
-        maybeUpdateCurrentMetadata();
-        // force recreate notification to ensure seek bar is shown when preparation finishes
-        NotificationUtil.getInstance().createNotificationIfNeededAndUpdate(this, true);
+    public void onEvents(@NonNull final com.google.android.exoplayer2.Player player,
```
Comment (litetex) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:2490]: Still waiting for the documentation what is happening here...

File: app/src/main/java/org/schabi/newpipe/player/Player.java [hunk 15]
```diff
@@ -2559,94 +2599,106 @@ public void onCues(@NonNull final List<Cue> cues) {
...
-     */
+     * @see com.google.android.exoplayer2.Player.Listener#onPlayerError(PlaybackException)
+     * */
+    @SuppressLint("SwitchIntDef")
     @Override
-    public void onPlayerError(@NonNull final ExoPlaybackException error) {
+    public void onPlayerError(@NonNull final PlaybackException error) {
         Log.e(TAG, "ExoPlayer - onPlayerError() called with:", error);
 
+        setRecovery();
```
Comment (litetex) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:2630]: Why is setRecovery called here (a comment in the code would be great)

File: app/src/main/java/org/schabi/newpipe/player/Player.java [hunk 16]
```diff
@@ -3410,12 +3412,15 @@ private void updateStreamRelatedViews() {
                 break;
 
             case VIDEO_STREAM:
-                if (info.getVideoStreams().size() + info.getVideoOnlyStreams().size() == 0) {
+                if (currentMetadata == null
+                        || !currentMetadata.getMaybeQuality().isPresent()
+                        || info.getVideoStreams().size() + info.getVideoOnlyStreams().size() == 0) {
```
Comment (litetex) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:3417]:
```suggestion
                        || (info.getVideoStreams().isEmpty() && info.getVideoOnlyStreams().isEmpty())) {
```

I would write it like this - better understandable IMHO

File: app/src/main/java/org/schabi/newpipe/player/Player.java [hunk 17]
```diff
@@ -3679,7 +3684,10 @@ private void onTextTracksChanged() {
         }
 
         // Normalize mismatching language strings
-        final String preferredLanguage = trackSelector.getPreferredTextLanguage();
+        final List<String> preferredLanguages =
+                trackSelector.getParameters().preferredTextLanguages;
+        final String preferredLanguage =
+                preferredLanguages.isEmpty() ? null : preferredLanguages.get(0);
```
Comment (litetex) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:3690]:
```suggestion
        final String preferredLanguage = trackSelector.getParameters().preferredTextLanguages
                .stream().findFirst().orElse(null);
```

Would write it like this (no additional variable required)

File: app/src/main/java/org/schabi/newpipe/player/mediaitem/MediaItemTag.java [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+/**
+ * Metadata container and accessor used by player internals.
+ *
+ * This interface ensures consistency of fetching metadata on each stream,
+ * which is encapsulated in a {@link MediaItem} and delivered via ExoPlayer's
+ * {@link Player.Listener} on event triggers to the downstream users.
+ **/
+public interface MediaItemTag {
+
+    List<Throwable> getErrors();
```
Comment (litetex) [inline app/src/main/java/org/schabi/newpipe/player/mediaitem/MediaItemTag.java:29]: I'm pretty sure that these should be Exceptions.
https://rules.sonarsource.com/java/RSPEC-1181

File: app/src/main/java/org/schabi/newpipe/player/mediasource/FailedMediaSource.java [hunk 19]
```diff
@@ -2,88 +2,162 @@
...
+     * while still provide a smooth playback experience. A duration lower than 1 second is
+     * not recommended, it may cause ExoPlayer to buffer for a while.
+     * */
+    public static final long SILENCE_DURATION_US = TimeUnit.SECONDS.toMicros(2);
+    public static final MediaPeriod SILENT_MEDIA = makeSilentMediaPeriod(SILENCE_DURATION_US);
+
     private final String TAG = "FailedMediaSource@" + Integer.toHexString(hashCode());
     private final PlayQueueItem playQueueItem;
-    private final FailedMediaSourceException error;
+    private final Throwable error;
```
Comment (litetex) [inline app/src/main/java/org/schabi/newpipe/player/mediasource/FailedMediaSource.java:39]: https://rules.sonarsource.com/java/RSPEC-1181

File: app/src/main/java/org/schabi/newpipe/player/Player.java [hunk 20]
```diff
@@ -2984,10 +3010,10 @@ public void saveStreamProgressState() {
     }
 
     public void saveStreamProgressStateCompleted() {
-        if (currentMetadata != null) {
+        getCurrentStreamInfo().ifPresent(info -> {
             // current stream has ended, so the progress is its duration (+1 to overcome rounding)
-            saveStreamProgressState((currentMetadata.getMetadata().getDuration() + 1) * 1000);
-        }
+            saveStreamProgressState((info.getDuration() + 1) * 1000);
+        });
```
Comment (litetex) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:3016]: https://sonarcloud.io/project/issues?resolved=false&types=CODE_SMELL&pullRequest=8020&id=TeamNewPipe_NewPipe&open=AX9__pDcgqUh_HHl9bWr

https://rules.sonarsource.com/java/RSPEC-1602

**Review by litetex — CHANGES_REQUESTED** [review]
Some minor things I noticed:

File: app/src/main/java/org/schabi/newpipe/player/Player.java [hunk 21]
```diff
@@ -2460,27 +2484,51 @@ private void setMuteButton(@NonNull final ImageButton button, final boolean isMu
     //////////////////////////////////////////////////////////////////////////*/
     //region ExoPlayer listeners (that didn't fit in other categories)
 
+    /**
+     * <p>Listens for event or state changes on ExoPlayer. When any event happens, we check for
```
Comment (litetex) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:2488]: Shouldn't the <p> be also closed somewhere?

File: app/src/main/java/org/schabi/newpipe/player/Player.java [hunk 22]
```diff
@@ -2559,94 +2610,100 @@ public void onCues(@NonNull final List<Cue> cues) {
...
+     * For any error above that is <b>not</b> explicitly <b>catchable</b>, the player will
+     * create a notification so users are aware.
      * </ul>
-     *
-     * @see #processSourceError(IOException)
-     * @see com.google.android.exoplayer2.Player.Listener#onPlayerError(ExoPlaybackException)
-     */
+     * @see com.google.android.exoplayer2.Player.Listener#onPlayerError(PlaybackException)
+     * */
+    @SuppressLint("SwitchIntDef")
```
Comment (litetex) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:2637]: A reason for the suppression would be good 😄

e.g.
```java
@SuppressLint("SwitchIntDef") // My cat doesn't like it,. No, seriously write a real reason here ;)
```

File: app/src/main/java/org/schabi/newpipe/player/mediasource/FailedMediaSource.java [hunk 23]
```diff
@@ -2,88 +2,162 @@
...
+        return new FailedMediaSource(playQueueItem, error,
+                System.currentTimeMillis() + retryWaitMillis);
     }
 
     public PlayQueueItem getStream() {
         return playQueueItem;
     }
 
-    public FailedMediaSourceException getError() {
+    public Throwable getError() {
```
Comment (litetex) [inline app/src/main/java/org/schabi/newpipe/player/mediasource/FailedMediaSource.java:81]: Shouldn't that be also an  ``Exception`` here?

File: app/src/main/java/org/schabi/newpipe/player/playback/MediaSourceManager.java [hunk 24]
```diff
@@ -417,20 +423,27 @@ private void maybeLoadItem(@NonNull final PlayQueueItem item) {
...
-            return new LoadedMediaSource(source, stream, expiration);
-        }).onErrorReturn(throwable -> new FailedMediaSource(stream,
-                new StreamInfoLoadException(throwable)));
+            return new LoadedMediaSource(source, tag, stream, expiration);
+        }).onErrorReturn(throwable -> {
+            if (throwable instanceof ExtractionException) {
+                return FailedMediaSource.of(stream, new StreamInfoLoadException(throwable));
+            }
+            return FailedMediaSource
+                    .of(stream, new Exception(throwable), /*immediatelyRetryable=*/0L);
```
Comment (litetex) [inline app/src/main/java/org/schabi/newpipe/player/playback/MediaSourceManager.java:445]: ``/*immediatelyRetryable=*/0L``
Couldn't that lead to a deadlock/loop?

**Review by litetex — CHANGES_REQUESTED** [review]
Nearly perfect, only minor things:

**Review by litetex — APPROVED** [review]
LGTM

Found one problem while testing (however this is likely unrelated): When enqueuing a video it gets enqueued 2x.

Otherwise it looks like it' working fine 👍

**Comment by karyogamy** [comment]
Other issues that should be fixed in tandem:
- Video progress resets during player switch: #7916
- Video Surface Released: #6989

**Comment by TiA4f8R** [comment]
It seems I found a regression with this PR: use the Play all button on the [Creative Commons channel](https://www.youtube.com/c/CreativeCommonsVids/videos), open the video with Stacy Allison-Cassin, enable autogenerated subtitles, then switch to the next video. Wait for lyrics and you will notice that English subtitles are enabled, even if subtitles are marked as disabled by the player.

**Comment by karyogamy** [comment]
Confirmed - can reproduce by queuing any 2 videos, first with a manual subtitle (e.g. english) and the second with only auto-generated subtitle of the previous language (e.g. english auto-generated) and the rest is the same as @TiA4f8R describes. 

I suspect this is because of how ExoPlayer compare language names, we can probably fix this by updating the selected ExoPlayer subtitle track according to the selected caption dropdown on metadata update. Will take a look when next available.

File: app/src/main/java/org/schabi/newpipe/player/Player.java [hunk 25]
```diff
@@ -2460,27 +2484,51 @@ private void setMuteButton(@NonNull final ImageButton button, final boolean isMu
...
-    public void onTimelineChanged(@NonNull final Timeline timeline, final int reason) {
-        if (DEBUG) {
-            Log.d(TAG, "ExoPlayer - onTimelineChanged() called with "
-                    + "timeline size = [" + timeline.getWindowCount() + "], "
-                    + "reason = [" + reason + "]");
-        }
-
-        maybeUpdateCurrentMetadata();
-        // force recreate notification to ensure seek bar is shown when preparation finishes
-        NotificationUtil.getInstance().createNotificationIfNeededAndUpdate(this, true);
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:2473]: Why was this removed @karyogamy?

File: app/src/main/java/org/schabi/newpipe/player/Player.java [hunk 26]
```diff
@@ -4519,9 +4538,13 @@ private void setupVideoSurface() {
...
             final Surface surface = binding.surfaceView.getHolder().getSurface();
-            // initially set the surface manually otherwise
-            // onRenderedFirstFrame() will not be called
-            simpleExoPlayer.setVideoSurface(surface);
+            // ensure player is using an unreleased surface, which the surfaceView might not be
+            // when starting playback on background or during player switching
+            if (surface.isValid()) {
+                // initially set the surface manually otherwise
+                // onRenderedFirstFrame() will not be called
+                simpleExoPlayer.setVideoSurface(surface);
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:4546]: Could this be replaced by `simpleExoPlayer.setVideoSurfaceHolder(binding.surfaceView.getHolder());`, or would there be issues? After some testing in #8170 I couldn't notice any difference.

File: app/src/main/java/org/schabi/newpipe/player/Player.java [hunk 27]
```diff
@@ -3043,48 +3071,29 @@ private void onMetadataChanged(@NonNull final MediaSourceTag tag) {
...
-        maybeAutoQueueNextStream(metadata);
-
-        if (currentMetadata == metadata) {
-            return;
-        }
-        currentMetadata = metadata;
-        onMetadataChanged(metadata);
+        maybeAutoQueueNextStream(streamInfo);
+        onMetadataChanged(streamInfo);
+        NotificationUtil.getInstance().createNotificationIfNeededAndUpdate(this, true);
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:3081]: Also, why does this now have `forceRecreate = true`? Also, it seems like you are calling `createNotificationIfNeededAndUpdate` twice when `updateMetadataWith` is called, since there is a `createNotificationIfNeededAndUpdate(this, false);` also inside `onMetadataChanged`. And as you can see that one has `forceRecreate = false`.

File: app/src/main/java/org/schabi/newpipe/player/Player.java (hunk 25 above)
Comment (karyogamy) [reply app/src/main/java/org/schabi/newpipe/player/Player.java:2473]: Updates still happen on timeline changes. This method, and other calls to `maybeUpdateCurrentMetadata` are now conslidated into `onEvents` which is triggered on any ExoPlayer change instead of scattering them into different callbacks.

File: app/src/main/java/org/schabi/newpipe/player/Player.java (hunk 27 above)
Comment (karyogamy) [reply app/src/main/java/org/schabi/newpipe/player/Player.java:3081]: Oversight on my part. `createNotificationIfNeededAndUpdate ` in `onMetadataChanged` should be removed.

File: app/src/main/java/org/schabi/newpipe/player/Player.java (hunk 26 above)
Comment (karyogamy) [reply app/src/main/java/org/schabi/newpipe/player/Player.java:4546]: Absolutely, that is the better solution. `setVideoSurfaceHolder` is doing the same null checks here: https://github.com/google/ExoPlayer/blob/release-v2/library/core/src/main/java/com/google/android/exoplayer2/ExoPlayerImpl.java#L1262

**Comment by karyogamy** [comment]
Sorry about that folks, got too used to Stash PR workflow and didn't realize on Github all my review replies stayed in `Pending` and never got out =\
